modules/system/applications.nix

```diff
@@ -37,11 +42,30 @@ in
         rm ~/Applications/'Nix Apps'
       fi
 
-      if [ ! -e '/Applications/Nix Apps' ] \
-         || ourLink '/Applications/Nix Apps'; then
-        ln -sfn ${cfg.build.applications}/Applications '/Applications/Nix Apps'
+      ourfolder () {
+        local marker
+        marker=$(xattr -p org.nix-darwin "$1" 2>/dev/null)
+        [ "$marker" = 'generated' ]
+      }
+
+      targetFolder='/Applications/Nix Apps'
+
+      # Clean up for links created at the old location in /Applications
+      if [ -e "$targetFolder" ] && ourLink "$targetFolder"; then
+        rm "$targetFolder"
+      fi
+
+      if [ -e "$targetFolder" ] && ! ourFolder "$targetFolder"; then
+        echo "warning: $targetFolder is not owned by nix-darwin, skipping App linking..." >&2
       else
-        echo "warning: /Applications/Nix Apps is not owned by nix-darwin, skipping App linking..." >&2
+        # create and mark folder
+        mkdir -p "$targetFolder"
+        xattr -w org.nix-darwin generated "$targetFolder"
+
+        # sync applications over resolving the first symlinks, but retaining all internal ones
+        find ${cfg.build.applications}/Applications -maxdepth 1 -type l -exec rsync -avL --delete {} "$targetFolder" \;
```

Comment:
`rsync` needs to use `--checksum`. Otherwise, it only compares based on mtime and size. The former will (or should) always be zero in the store, which means it effectively only compares by size. If a new file has the same size as the old one, the old one will be left in place, which can cause problems.

Reply:
We should ideally look at the store path the app is from to avoid the expense of recomputing the hash on activation.

@Samasaur1 Will your files module work make this kind of thing easier? If so it might be better to land that before this PR so it can utilize the logic.

Reply:
How would we know the store path of the app if it’s been copied out of the store? It’s probably not worth the extra complexity of trying to track it unless we use something like @Samasaur1’s PR. One issue is we won’t know what files to copy until activation. Does his PR have activation helpers this one could use?

Reply:
@emilazy This diff is slightly outdated, I since understood from @reckenrode that we just need one rsync to achieve what we want. Maybe I misunderstood you, but that should mean that Applications change only when they actually change on disk. That _should_ prevent hash recompilation I think?

Reply:
> How would we know the store path of the app if it’s been copied out of the store?

Well, we’re already writing `xattr`s, right? But I agree that it might make sense to land this without and port it to the file mechanism later. I just worry about the speed of checksumming, since Mac apps can be pretty big.

> @emilazy This diff is slightly outdated, I since understood from @reckenrode that we just need one rsync to achieve what we want. Maybe I misunderstood you, but that should mean that Applications change only when they actually change on disk. That _should_ prevent hash recompilation I think?

I think that currently, `rsync` will always hash every single file in `${cfg.build.applications}/Applications` and `$targetFolder` to determine whether any of them changed.

Reply:
That is correct. The filesystem could provide an optimization for this, as apfs does know if anything in a subtree has changed. I do not know wether `rsync` has any way to take advantage of that though.

Reply:
In my tests, this takes around 2.5 Seconds of wall time to copy around 800mb of data (ut1999 for me) initially and around 1.3s when source and target are identical. That is not nothing, but also not too bad.